trio/_threads.py

```diff
@@ -377,6 +379,9 @@ def from_thread_run(afn, *args, trio_token=None):
           "foreign" thread, spawned using some other framework, and still want
           to enter Trio.
     """
+    if not inspect.iscoroutinefunction(afn):
```

Comment:
Unfortunately, identifying async functions is rather subtler than just `inspect.iscoroutinefunction`, because async functions can be wrapped in a way that doesn't satisfy that test. (For example, Trio functions decorated with `@enable_ki_protection` are implemented using a wrapper that looks synchronous. `functools.partial` objects don't even satisfy `inspect.isfunction`, let alone `inspect.iscoroutinefunction`. All of these are perfectly valid to pass as an "async function" as long as the function being wrapped is async.)

The best way to see if something is an async function is to call it without `await` and use `inspect.iscoroutine` on the result. Of course, if it wasn't an async function, you just executed it, so it's best to do this only in the context where you were planning to call it anyway. In this case, that would be inside `unprotected_afn` below.

If you're not willing to call the function, you have to get a lot more special-cased to see if it's async or not. You can look inside the `.func` attribute of a `functools.partial` object, follow the `.__wrapped__` attributes generally added by function wrappers, etc, but there's no approach that will definitely succeed for any async callable.